sdk/include/reactos/debug.h

```diff
@@ -58,6 +61,77 @@ RtlAssert(
     _In_opt_z_ PCHAR Message
 );
 
+#elif defined(_WIN32)
+/* Win32 User-Mode */
+
+#ifndef NO_STRSAFE
+#include <strsafe.h>
+#endif
+
+#ifndef DEBUG_BUFSIZE
+#define DEBUG_BUFSIZE 512
+#endif
+
+static inline
+ULONG
+__cdecl
+W32DbgPrint(_In_z_ _Printf_format_string_ PCSTR Format, ...)
+{
+    va_list va;
+    char szBuff[DEBUG_BUFSIZE];
+
+    va_start(va, Format);
+
+#ifdef NO_STRSAFE
+    _vsnprintf(szBuff, _countof(szBuff), Format, va);
+#else
+    StringCchVPrintfA(szBuff, _countof(szBuff), Format, va);
+#endif
+
+    OutputDebugStringA(szBuff);
+    va_end(va);
+
+    return 0; /* Return STATUS_SUCCESS, since we are supposed to mimic DbgPrint */
+}
+
+#define DbgPrint(Format, ...) W32DbgPrint(Format, __VA_ARGS__)
+#define DbgPrintEx(ComponentId, Level, Format, ...) W32DbgPrint(Format, __VA_ARGS__)
+
+static inline
+__analysis_noreturn
+VOID
+NTAPI
+W32Assert(
+    _In_ PVOID FailedAssertion,
+    _In_ PVOID FileName,
+    _In_ ULONG LineNumber,
+    _In_opt_z_ PCHAR Message)
+{
+    INT id;
+    char szBuff[DEBUG_BUFSIZE];
+
+#ifdef NO_STRSAFE
+    _snprintf(szBuff, _countof(szBuff),
+#else
+    StringCchPrintfA(szBuff, _countof(szBuff),
+#endif
+                     "File '%s',\nLine %ld:\n\n%s\n\n%s",
```

Comment:
Is this assertion message alright?

Reply:
For information, here is what the ntdll RtlAssert calls when displaying the assertion:
```c
    /* Print the assertion */
    DbgPrint("\n*** Assertion failed: %s%s\n"
             "***   Source File: %s, line %lu\n\n",
             Message != NULL ? Message : "",
             (PSTR)FailedAssertion,
             (PSTR)FileName,
             LineNumber);
```
(note the `Message` NULL check)

Reply:
You can also observe that the win32 CRT has equivalent functions, and you may be inspired by its assertion message:
https://git.reactos.org/?p=reactos.git;a=blob;f=sdk/lib/crt/misc/dbgrpt.cpp;hb=47f6745bcdfb833b0311e052f04e6cd5634660eb#l79
(Note: these can only be used in a win32 user-mode app that explicitly uses the CRT.)